rust/otap-dataflow/crates/engine/src/effect_handler.rs

```diff
@@ -69,4 +69,49 @@ impl EffectHandlerCore {
 
         TcpListener::from_std(sock.into()).map_err(err)
     }
+
+    /// Creates a non-blocking UDP socket on the given address with socket options defined by the
+    /// pipeline engine implementation. It's important for receiver implementer to create UDP
+    /// sockets via this method to ensure the scalability and the serviceability of the pipeline.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`Error::IoError`] if any step in the process fails.
+    ///
+    /// ToDo: return a std::net::UdpSocket instead of a tokio::net::UdpSocket to avoid leaking our current dependency on Tokio.
+    pub(crate) fn udp_socket<PData>(
+        &self,
+        addr: SocketAddr,
+        receiver_name: impl Into<Cow<'static, str>>,
+    ) -> Result<UdpSocket, Error<PData>> {
+        let node_name: Cow<'static, str> = receiver_name.into();
+        // Helper closure to convert errors.
+        let err = |error: std::io::Error| Error::IoError {
+            node: node_name.clone(),
+            error,
+        };
```

Comment:
nit: I would name this `to_err`

Curiosity question: I wonder if `thiserror` is meant to help us not write this sort of thing, but maybe it doesn't. I feel like the `snafu` crate has a way to accomplish the insertion of the node name using a `.with_context(|| ...)` expression to add the custom fields of the structured object while propagating a source error. Since all of the errors here are IO errors and since the only reason receiver_name is passed here is to form an error, I would be inclined to let it return the io Error and allow the caller to add the node_name context.